File: web/src/firebase-sync.js
import {
  ref, set, get, remove, onValue, off,
  serverTimestamp, query, orderByChild,
  onDisconnect
} from 'firebase/database';
import { getFirebaseDb, ensureFirebaseAuth } from './firebase-client.js';

function roomRef(roomCode) {
  return ref(getFirebaseDb(), `movsync/rooms/${roomCode.toUpperCase()}`);
}

function childRef(roomCode, path) {
  return ref(getFirebaseDb(), `movsync/rooms/${roomCode.toUpperCase()}/${path}`);
}

export async function createRoomNode(roomCode) {
  await set(childRef(roomCode, 'createdAt'), serverTimestamp());
}

export async function setRoomStarted(roomCode) {
  await set(childRef(roomCode, 'started'), true);
}

export function observeRoomStarted(roomCode, callback) {
  const r = childRef(roomCode, 'started');
  const unsubscribe = onValue(r, (snapshot) => {
    callback(snapshot.val() === true);
  });
  return () => off(r);
}

export async function setAllowControls(roomCode, allow) {
  await set(childRef(roomCode, 'allowControls'), allow);
}

export function observeAllowControls(roomCode, callback) {
  const r = childRef(roomCode, 'allowControls');
  onValue(r, (snapshot) => {
    if (!snapshot.exists()) {
      callback(true);
    } else {
      callback(snapshot.val() === true);
    }
  });
  return () => off(r);
}

export async function writeSyncCommand(roomCode, state) {
  const payload = {
    command: state.command,
    position: state.position,
    speed: state.speed,
    timestamp: serverTimestamp(),
    senderId: state.senderId,
  };
  await set(childRef(roomCode, 'sync'), payload);
}

export async function writeHeartbeat(roomCode, state) {
  const payload = {
    position: state.position,
    isPlaying: state.isPlaying,
    timestamp: serverTimestamp(),
  };
  await set(childRef(roomCode, 'heartbeat'), payload);
}

export async function getHeartbeatOnce(roomCode) {
  const snapshot = await get(childRef(roomCode, 'heartbeat'));
  if (!snapshot.exists()) return null;
  const val = snapshot.val();
  return {
    position: val.position || 0,
    isPlaying: val.isPlaying || false,
    timestamp: val.timestamp || 0,
  };
}

export function listenToSync(roomCode, callback) {
  const r = childRef(roomCode, 'sync');
  onValue(r, (snapshot) => {
    if (!snapshot.exists()) return;
    const val = snapshot.val();
    callback({
      command: val.command || '',
      position: val.position || 0,
      speed: val.speed != null ? val.speed : 1,
      timestamp: val.timestamp || 0,
      senderId: val.senderId || '',
    });
  });
  return () => off(r);
}

export function listenToHeartbeat(roomCode, callback) {
  const r = childRef(roomCode, 'heartbeat');
  onValue(r, (snapshot) => {
    if (!snapshot.exists()) return;
    const val = snapshot.val();
    callback({
      position: val.position || 0,
      isPlaying: val.isPlaying || false,
      timestamp: val.timestamp || 0,
    });
  });
  return () => off(r);
}

export async function trackPresence(roomCode, userId, displayName, isHost, verified) {
  await ensureFirebaseAuth();

  const presRef = childRef(roomCode, `presence/${userId}`);
  const payload = {
    displayName,
    isHost,
    online: true,
    verified,
    updatedAt: serverTimestamp(),
  };
  await set(presRef, payload);

  // Set online to false on disconnect
  const onlineRef = childRef(roomCode, `presence/${userId}/online`);
  const disc = onDisconnect(onlineRef);
  await disc.set(false);

  // If host, remove entire room node on disconnect
  if (isHost) {
    const roomDisc = onDisconnect(roomRef(roomCode));
    await roomDisc.remove();
  }
}

export async function clearPresence(roomCode, userId) {
  await remove(childRef(roomCode, `presence/${userId}`));
}

export function observePresence(roomCode, callback) {
  const r = childRef(roomCode, 'presence');
  onValue(r, (snapshot) => {
    if (!snapshot.exists()) {
      callback([]);
      return;
    }
    const users = [];
    snapshot.forEach((child) => {
      const val = child.val();
      users.push({
        userId: child.key,
        displayName: val.displayName || '',
        isHost: val.isHost || false,
        online: val.online || false,
        verified: val.verified || false,
      });
    });
    callback(users);
  });
  return () => off(r);
}

export async function sendMessage(roomCode, message) {
  const payload = {
    senderId: message.senderId,
    senderName: message.senderName,
    message: message.message,
    timestamp: serverTimestamp(),
  };
  await set(childRef(roomCode, `chat/${message.messageId}`), payload);
}

export function observeChatMessages(roomCode, callback) {
  const r = query(childRef(roomCode, 'chat'), orderByChild('timestamp'));
  onValue(r, (snapshot) => {
    if (!snapshot.exists()) {
      callback([]);
      return;
    }
    const messages = [];
    snapshot.forEach((child) => {
      const val = child.val();
      messages.push({
        messageId: child.key,
        senderId: val.senderId || '',
        senderName: val.senderName || '',
        message: val.message || '',
        timestamp: val.timestamp || 0,
      });
    });
    callback(messages);
  });
  return () => off(r);
}

export function removeAllListeners(roomCode) {
  off(childRef(roomCode, 'sync'));
  off(childRef(roomCode, 'heartbeat'));
  off(childRef(roomCode, 'presence'));
  off(childRef(roomCode, 'chat'));
  off(childRef(roomCode, 'started'));
  off(childRef(roomCode, 'allowControls'));
}

File: app/src/main/java/com/nityam/movsync/data/firebase/FirebaseSync.kt
package com.nityam.movsync.data.firebase

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ServerValue
import com.google.firebase.database.ValueEventListener
import com.nityam.movsync.data.model.HeartbeatState
import com.nityam.movsync.data.model.PresenceUser
import com.nityam.movsync.data.model.SyncState
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

class FirebaseSync(
    private val database: FirebaseDatabase,
    private val firebaseAuth: FirebaseAuth
) {
    private var serverTimeOffset: Long = 0L

    init {
        database.getReference(".info/serverTimeOffset").addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                serverTimeOffset = snapshot.getValue(Long::class.java) ?: 0L
            }
            override fun onCancelled(error: DatabaseError) = Unit
        })
    }

    fun getEstimatedServerTime(): Long {
        return System.currentTimeMillis() + serverTimeOffset
    }
    private fun roomRef(roomCode: String) = database.reference
        .child("movsync")
        .child("rooms")
        .child(roomCode.uppercase())

    suspend fun createRoomNode(roomCode: String) {
        roomRef(roomCode).child("createdAt").setValue(ServerValue.TIMESTAMP).await()
    }

    suspend fun setRoomStarted(roomCode: String) {
        roomRef(roomCode).child("started").setValue(true).await()
    }

    fun observeRoomStarted(roomCode: String): Flow<Boolean> = callbackFlow {
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                trySend(snapshot.getValue(Boolean::class.java) ?: false)
            }
            override fun onCancelled(error: DatabaseError) {
                close(error.toException())
            }
        }
        val ref = roomRef(roomCode).child("started")
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    suspend fun setAllowControls(roomCode: String, allow: Boolean) {
        roomRef(roomCode).child("allowControls").setValue(allow).await()
    }

    fun observeAllowControls(roomCode: String): Flow<Boolean> = callbackFlow {
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                trySend(snapshot.getValue(Boolean::class.java) ?: true)
            }
            override fun onCancelled(error: DatabaseError) {
                close(error.toException())
            }
        }
        val ref = roomRef(roomCode).child("allowControls")
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    suspend fun setVoiceActive(roomCode: String, active: Boolean) {
        roomRef(roomCode).child("voiceActive").setValue(active).await()
    }

    fun observeVoiceActive(roomCode: String): Flow<Boolean> = callbackFlow {
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                trySend(snapshot.getValue(Boolean::class.java) ?: false)
            }

            override fun onCancelled(error: DatabaseError) {
                close(error.toException())
            }
        }
        val ref = roomRef(roomCode).child("voiceActive")
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    suspend fun writeSyncCommand(roomCode: String, state: SyncState) {
        val payload = mapOf(
            "command" to state.command,
            "position" to state.position,
            "speed" to state.speed,
            "timestamp" to ServerValue.TIMESTAMP,
            "senderId" to state.senderId
        )
        roomRef(roomCode).child("sync").setValue(payload).await()
    }

    suspend fun writeHeartbeat(roomCode: String, state: HeartbeatState) {
        val payload = mapOf(
            "position" to state.position,
            "isPlaying" to state.isPlaying,
            "timestamp" to ServerValue.TIMESTAMP
        )
        roomRef(roomCode).child("heartbeat").setValue(payload).await()
    }

    suspend fun trackPresence(
        roomCode: String,
        userId: String,
        displayName: String,
        isHost: Boolean,
        verified: Boolean
    ) {
        if (firebaseAuth.currentUser == null) {
            firebaseAuth.signInAnonymously().await()
        }

        val presenceRef = roomRef(roomCode).child("presence").child(userId)
        val payload = mapOf(
            "displayName" to displayName,
            "isHost" to isHost,
            "online" to true,
            "verified" to verified,
            "updatedAt" to ServerValue.TIMESTAMP
        )
        presenceRef.setValue(payload).await()
        presenceRef.child("online").onDisconnect().setValue(false)
    }

    suspend fun setPresenceVerified(roomCode: String, userId: String, verified: Boolean = true) {
        if (firebaseAuth.currentUser == null) {
            firebaseAuth.signInAnonymously().await()
        }
        val presenceRef = roomRef(roomCode).child("presence").child(userId)
        presenceRef.child("verified").setValue(verified).await()
        presenceRef.child("updatedAt").setValue(ServerValue.TIMESTAMP).await()
    }

    suspend fun clearPresence(roomCode: String, userId: String) {
        roomRef(roomCode).child("presence").child(userId).removeValue().await()
    }

    suspend fun getHeartbeatOnce(roomCode: String): HeartbeatState? {
        return roomRef(roomCode).child("heartbeat").get().await().toHeartbeatStateOrNull()
    }

    fun listenToSync(roomCode: String, onState: (SyncState) -> Unit): ValueEventListener {
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                snapshot.toSyncStateOrNull()?.let(onState)
            }

            override fun onCancelled(error: DatabaseError) = Unit
        }
        roomRef(roomCode).child("sync").addValueEventListener(listener)
        return listener
    }

    fun listenToHeartbeat(roomCode: String, onState: (HeartbeatState) -> Unit): ValueEventListener {
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                snapshot.toHeartbeatStateOrNull()?.let(onState)
            }

            override fun onCancelled(error: DatabaseError) = Unit
        }
        roomRef(roomCode).child("heartbeat").addValueEventListener(listener)
        return listener
    }

    fun presence(roomCode: String): Flow<List<PresenceUser>> = callbackFlow {
        val ref = roomRef(roomCode).child("presence")
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                trySend(snapshot.children.map { it.toPresenceUser() })
            }

            override fun onCancelled(error: DatabaseError) {
                close(error.toException())
            }
        }
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    fun removeSyncListener(roomCode: String, listener: ValueEventListener) {
        roomRef(roomCode).child("sync").removeEventListener(listener)
    }

    fun removeHeartbeatListener(roomCode: String, listener: ValueEventListener) {
        roomRef(roomCode).child("heartbeat").removeEventListener(listener)
    }

    suspend fun sendMessage(roomCode: String, message: com.nityam.movsync.data.model.ChatMessage) {
        val payload = mapOf(
            "senderId" to message.senderId,
            "senderName" to message.senderName,
            "message" to message.message,
            "timestamp" to ServerValue.TIMESTAMP
        )
        roomRef(roomCode).child("chat").child(message.messageId).setValue(payload).await()
    }

    fun observeChatMessages(roomCode: String): Flow<List<com.nityam.movsync.data.model.ChatMessage>> = callbackFlow {
        val ref = roomRef(roomCode).child("chat").orderByChild("timestamp")
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val messages = snapshot.children.mapNotNull { it.toChatMessageOrNull() }
                trySend(messages)
            }

            override fun onCancelled(error: DatabaseError) {
                close(error.toException())
            }
        }
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    private fun DataSnapshot.toSyncStateOrNull(): SyncState? {
        if (!exists()) return null
        return SyncState(
            command = child("command").getValue(String::class.java).orEmpty(),
            position = child("position").getValue(Long::class.java) ?: 0L,
            speed = child("speed").getValue(Double::class.java)?.toFloat()
                ?: child("speed").getValue(Float::class.java)
                ?: 1f,
            timestamp = child("timestamp").getValue(Long::class.java) ?: 0L,
            senderId = child("senderId").getValue(String::class.java).orEmpty()
        )
    }

    private fun DataSnapshot.toHeartbeatStateOrNull(): HeartbeatState? {
        if (!exists()) return null
        return HeartbeatState(
            position = child("position").getValue(Long::class.java) ?: 0L,
            isPlaying = child("isPlaying").getValue(Boolean::class.java) ?: false,
            timestamp = child("timestamp").getValue(Long::class.java) ?: 0L
        )
    }

    private fun DataSnapshot.toPresenceUser(): PresenceUser {
        return PresenceUser(
            userId = key.orEmpty(),
            displayName = child("displayName").getValue(String::class.java).orEmpty(),
            isHost = child("isHost").getValue(Boolean::class.java) ?: false,
            online = child("online").getValue(Boolean::class.java) ?: false,
            verified = child("verified").getValue(Boolean::class.java) ?: false
        )
    }

    private fun DataSnapshot.toChatMessageOrNull(): com.nityam.movsync.data.model.ChatMessage? {
        if (!exists()) return null
        return com.nityam.movsync.data.model.ChatMessage(
            messageId = key.orEmpty(),
            senderId = child("senderId").getValue(String::class.java).orEmpty(),
            senderName = child("senderName").getValue(String::class.java).orEmpty(),
            message = child("message").getValue(String::class.java).orEmpty(),
            timestamp = child("timestamp").getValue(Long::class.java) ?: 0L
        )
    }
}
